### PR TITLE
EID-1293 - Create HashFactory to hash certain attributes of the response

### PR DESCRIPTION
**Pull request by JHjava**
- This HashFactory will be used in 2 scenarios 
 Hub - To Hash the attributes of responses that are going back to the proxy node
 Proxy Node - To Hash the attributes of the response from the Hub
- Check the Date format matches YYYY-MM-DD and throw if not

### CODE REVIEW

File: saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+
+        String toHash = MessageFormat.format("{0},{1},{2},{3},{4}",
+                pid, firstName, middlename, lastName, dateOfBirth);
+
+        try {
+            messageDigest.update(toHash.getBytes("UTF-8"));
+        } catch (UnsupportedEncodingException e) {
```
Comment (tomrosier) [inline saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java]: Make this catch the `NoSuchAlgorithmException` to. and stick both
```
messageDigest = MessageDigest.getInstance(JCAConstants.DIGEST_SHA256);
```
and 
```
String toHash = MessageFormat.format("{0},{1},{2},{3},{4}",
          pid, firstName, middlename, lastName, dateOfBirth);
```
In the same try catch block.

File: saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            throw new RuntimeException(e);
```
Comment (tomrosier) [inline saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java]: `RuntimeException` is a bit generic, can we have a specific exception handler for this, to be consistant with the rest

File: saml-utils/src/test/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    public ExpectedException exception = ExpectedException.none();
+
+    @Test
+    public void shouldCreateHashOfResponseAttributes() {
+        String hashedResponse = hashResponseDetails("pid", "jim", "bob", "joe", "1989-10-10");
+
+        assertEquals("Hashes do not match up", hashedResponse, "437256f80cf561b2fa195bd1ed4293d4432d1c83090558bf3dd568c272943fd8");
+    }
+
+    @Test
```
Comment (tomrosier) [inline saml-utils/src/test/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactoryTest.java]: `@Test(expected = RuntimeException.class)` may be nicer 🙂

Comment (JHjava) [reply saml-utils/src/test/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactoryTest.java]: There's 3 RuntimeException that are being thrown in the class, so this way I can guarantee that the correct one is being thrown.

File: saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java (hunk 1 above)
Comment (JHjava) [reply saml-utils/src/main/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactory.java]: Yep, can do. I'll stick them in the same block but keep it as a Runtime exception which is thrown.

File: saml-utils/src/test/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactoryTest.java (hunk 3 above)
Comment (JHjava) [reply saml-utils/src/test/java/uk/gov/ida/saml/core/transformers/ResponseAttributesHashFactoryTest.java]: *2 now

**Pull request merged**